### DISCUSSION TITLE
Eloquent casting. How to use a array of abstract dto?

### DISCUSSION THREAD

**Discussion by RomainMazB** [spatie/laravel-data#705] [Q&A] [answered]
Hi!

I would like to use [the abstract dto feature](https://spatie.be/docs/laravel-data/v4/advanced-usage/eloquent-casting#content-abstract-data-objects) within an array with the [nesting collection feature](https://spatie.be/docs/laravel-data/v4/as-a-data-transfer-object/nesting#content-collections-of-data-objects).

I've created what looks to be the correct structure:
```php
// The abstract class
abstract class ActionConditionBase extends Data
{}

// The instantiatable class
class HasExecutedActionCondition extends ActionConditionBase implements NPCCondition
{
    public function __construct(
        public ActionDescription $actionDescription
    )
    {}
}

// The DTO storing the array of abstract/generic ActionConditionBase
class NPCAction extends Data
{
    public function __construct(
        public string $identifier,
        public int $times = 1,
        /** @var array<ActionConditionBase> */
        public ?array $conditions = []
    )
    {}
}
```

If I try to store this inside the database, refresh the model and try to dd it:
```php
$npc = NPC::factory()
    ->create([
        'data' => new \App\Classes\NPCAction('talk', conditions: [
            new \App\Classes\HasExecutedActionCondition(
                new \App\Classes\ActionDescription($firstNpc, $actionName, 2))
        ])
    ]);
dd($npc->refresh()->data->conditions);
```
It fails in:
```sh
Error: Cannot instantiate abstract class App\Classes\ActionConditionBase
```

Is there a way to use an array of generic of should I use a custom transformer?

**Comment by RomainMazB** [comment]
I read https://github.com/spatie/laravel-data/discussions/407

Is that unchecked checkbox the feature I'm searching for 😢 ?
![image](https://github.com/spatie/laravel-data/assets/53976837/b5392427-8673-4126-ba55-1bdd43264755)

**Reply by rubenvanassche** [reply, marked as answer]
Yeah we do not support it at the moment since it requires a lot of work to get things like validation, partial includes and magical creation right

**Reply by RomainMazB** [reply]
Ok, thanks for your quick reply.

I'm using laravel-data just to provide a convenient way to store some dto as array in database and restore them.

With that in mind, could just say to me if a transformer could do this? or which section of the doc I should read?

**Reply by rubenvanassche** [reply]
I would just try some things, data is such a complicated machine of things going on. Even I don't completely understand it at certain times 😄
